sem/utils.py

```diff
@@ -330,3 +331,104 @@ def compute_sensitivity_analysis(
 #                  np.array(xarray.sel(**kwargs)).squeeze())
 #     interact(plot_line, **{k: v for k, v in param_ranges.items() if k != x_axis
 #                            and len(v) > 1})
+
+def parse_log_component(log_component, ns3_log_components=None):
+    """
+    Verifies if the log levels/log classes passed in the log_component
+    dictionary are valid and converts log levels to corresponding log classes.
+    Returns a dictionary with the valid components and log classes.
+    For example,
+    'level_debug' gets converted to 'warn|error|debug'
+
+    Args:
+        log_component (dict): a python dictionary with the
+            log_components (to enable) as the key and the log_levels
+            as the value. Log levels should be written in the same format as
+            the one specified by the ns-3 manual for the environment variable
+            NS_LOG.
+            Note: If any prefix is mentioned, it will be dropped and
+            prefix_all will always be appended.
+
+            For example,
+            log_component = {
+                'component1' : 'info',
+                'component2' : 'level_debug|info'
+            }
+        ns3_log_components (list): A list containing all the valid log
+            components supported by ns-3.
+    """
+    if not log_component:
+        return None
+
+    ret_dict = {}
+    log_level_list = ['error',
+                      'warn',
+                      'debug',
+                      'info',
+                      'function',
+                      'logic']
+    converter = {
+        'error': ['error'],
+        'warn': ['warn'],
+        'debug': ['debug'],
+        'info': ['info'],
+        'function': ['function'],
+        'logic': ['logic'],
+        'all': ['error', 'warn', 'debug', 'info', 'function', 'logic'],
+        'level_error': ['error'],
+        'level_warn': ['error', 'warn'],
+        'level_debug': ['error', 'warn', 'debug'],
+        'level_info': ['error', 'warn', 'debug', 'info'],
+        'level_function': ['error', 'warn', 'debug', 'info', 'function'],
+        'level_logic': ['error', 'warn', 'debug', 'info', 'function', 'logic'],
+        'level_all': ['error', 'warn', 'debug', 'info', 'function', 'logic'],
+        '**': ['error', 'warn', 'debug', 'info', 'function', 'logic'],
+        '*': None,
+        'prefix_func': None,
+        'prefix_time': None,
+        'prefix_node': None,
+        'prefix_level': None,
+        'prefix_all': None
+    }
+    for component, levels in log_component.items():
+        log_level_complete = set()
+        if (ns3_log_components is not None) and not (component not in ns3_log_components or component != '*'):
```

Comment:
dangerous condition: it's extremely complicate to read and make sure that it's correct (especially edge cases).
could you break it down a bit? or deconstruct it in different parts of the code using simpler conditions?

The following should be equivalent ([right?](https://en.wikipedia.org/wiki/De_Morgan%27s_laws))
```suggestion
        if (ns3_log_components is not None) and ((component == '*') and (component in ns3_log_components)):
```
I'm not sure if this is correct